src/features/simd.md

```diff
@@ -0,0 +1,179 @@
+
+```
+void multiply_arrays(int* out, int* in_a, int* in_b, int size) {
+
+    for (int i = 0; i < size; i++) {
+
+        out[i] = in_a[i] * in_b[i];
+
+    }
+
+}
+```
+
+Without passing the -msimd128 flag, the compiler emits this WebAssembly loop:
+
+```
+(loop
+
+    (i32.store
+
+        ... get address in `out` …
+
+        (i32.mul
+
+            (i32.load ... get address in `in_a` ...)
+
+            (i32.load ... get address in `in_b` ...)
+
+    ...
+
+)
+```
+
+But when the -msimd128 flag is used, the autovectorizer will turn this into code that includes this loop:
+
+```
+(loop
+
+    (v128.store align=4
+
+        ... get address in `out` ...
+
+        (i32x4.mul
+
+               (v128.load align=4 ... get address in `in_a` ...)
+
+               (v128.load align=4 ... get address in `in_b` ...)
+
+    ...
+
+    )
+
+)
+```
+
+As you can see, the loop body has the same structure but SIMD instructions are being used to load, multiply, and store four elements at a time inside the loop body.
+
+For finer grained control over the SIMD instructions generated by the compiler, include [this header file](https://github.com/emscripten-core/emscripten/blob/incoming/system/include/wasm_simd128.h), which defines a set of intrinsics. Intrinsics are special functions that, when called, will be turned by the compiler into the corresponding WebAssembly SIMD instructions, unless it can make further optimizations.
+
+As an example, here is the same function from before manually rewritten to use the SIMD intrinsics.
+
+```
+#include <wasm_simd128.h>
+
+void multiply_arrays(int* out, int* in_a, int* in_b, int size) {
+
+    for (int i = 0; i < size; i += 4) {
+
+        v128_t a = wasm_v128_load(&in_a[i]);
+
+        v128_t b = wasm_v128_load(&in_b[i]);
+
+        v128_t prod = wasm_i32x4_mul(a, b);
+
+        wasm_v128_store(&out[i], prod);
+
+    }
+
+}
+```
+
+This manually rewritten code assumes that the input and output arrays are aligned and do not alias and that size is a multiple of four. The autovectorizer cannot make these assumptions and has to generate extra code to handle the cases where they are not true, so hand-written SIMD code often ends up being smaller than autovectorized SIMD code.
+
+## Compelling Use Cases
+
+The WebAssembly SIMD proposal seeks to accelerate high compute applications like audio/video codecs, image processing applications, cryptographic applications, etc. Currently WebAssembly SIMD is experimentally supported in widely used open source projects like [Halide]([https://github.com/halide/Halide/blob/master/README_webassembly.md](https://github.com/halide/Halide/blob/master/README_webassembly.md)), [OpenCV.js]([https://docs.opencv.org/3.4/d5/d10/tutorial_js_root.html](https://docs.opencv.org/3.4/d5/d10/tutorial_js_root.html)), and [XNNPACK]([https://github.com/google/XNNPACK](https://github.com/google/XNNPACK)). [TODO: add links]
```

Comment:
These links don't look right: [name] ([url](url)).
Also, there is still the "TODO: add links".